### ISSUE TITLE
Warn if the className prop doesn't get used by 3rd party components

### ISSUE THREAD

**Issue by k15a** [styled-components/styled-components#349]
@vdanchenkov had an awesome idea at https://github.com/styled-components/styled-components/pull/268#issuecomment-269292621

> I have another idea of runtime warning that could be useful in development.
> 
> A lot of people misses the fact that 3-rd party components should accept and use className property. We could check that generated class name actually been used.
> 
> ```JS
> if (!document.querySelector('.' + className)) warn()
> ```
> Main concern is that there might be cases when you actually do not want to render the element with passed className. I have no examples of it, but if there are some we need to have a method to mute this warning.

Related issues: #309, #296, #265, #251, #141 and many more.

**Comment by mxstbr** [comment]
I feel like this would be a very positive change as this is by far the most often asked question. Let's do it!

**Comment by geelen** [comment]
This is a nice idea but that implementation suggestion is... kinda awkward. Could we not do something instead with `propTypes` so that if the component you pass doesn't even declare `className` as a prop we log a warning?

**Comment by vdanchenkov** [comment]
Yeah, it would be a lot simpler to implement, but I'm afraid there will be a lot of false positives. 

It would force users to start using propTypes to mute warnings. Sometimes you can't control it if it's third party component. Sometimes you don't want to.

**Comment by geelen** [comment]
Yeah I'm not sure it's valuable enough to warrant the noise. It would probably be best to expose some documentation around this issue instead. I've always wanted a "how does SC work?" "It just passes `className` as a property" area of the docs, which might solve this?

**Comment by k15a** [comment]
> For the styles to be applied, third-party components must attach the passed-in className prop to a DOM node.

There is already a note in the readme and people are not reading it :(

**Comment by mxstbr** [comment]
Yeah the docs are definitely not cutting it, we get asked this questions a lot even though it's in the README.

**Comment by k15a** [comment]
> Main concern is that there might be cases when you actually do not want to render the element with passed className. I have no examples of it, but if there are some we need to have a method to mute this warning.

This will probably be the case with #184.

**Comment by jpellizzari** [comment]
Hello! I recently ran into this issue and would like to help solve it. Does anyone have any thoughts on how this might be implemented?

My initial thoughts were to run `createElement` on the target component with a `className` prop passed in, and see what props are present:

```javascript
const createWarnNoClassName = (target: Target) => {
  const { props } = React.createElement(target, { className: 'foo' })

  if (!props.className) {
    console.warn('Whoops!')
  }
}
```
This seems like a heavy approach from a performance perspective. It could be conditionally executed on if `isTag(target)` is false to help reduce the footprint.

Anyone have any better ideas?

**Comment by probablyup** [comment]
I like this for dev only.

**Comment by James-E-Adams** [comment]
Definitely a fan of having a warning. 

Just spent 10-20 minutes trying to work out why my styles weren't getting added until it hit me that I'd forgotten to do this! 

Could we do it with an eslint rule? E.g.

```javascript
import React from 'react';
import styled from 'styled-components';

const sampleComponent= ({}) => {
  return (
    <div
    />
  );
};

const styledSampleComponent = styled(sampleComponent) 
```
should give a linting error such as:
`Cannot call styled on a component that does not accept a className prop`

**Comment by probablyup** [comment]
> Could we do it with an eslint rule

I strongly suspect yes! https://github.com/evcohen/eslint-plugin-jsx-a11y has a lot of rules that track usage of props you could probably borrow code from. We don't currently have an eslint plugin repo though, so perhaps you could own this and contribute it back to the community @James-E-Adams?

**Comment by James-E-Adams** [comment]
Yup, sure! Might take me a week or so before I get a chance to work on it though 🎏

**Comment by azz** [comment]
There's always the issue of people finding the ESLint rule _after_ finding they've missed a `className` propagation. At which time the rule will offer less value.

Another approach which could catch some cases and have fewer false positives is warning when `styled()` is called on a `Component` where:

```js
Component.propTypes && !Component.propTypes.className

**Comment by azz** [comment]
Also as an eslint rule you'll likely only be able to catch it when the component definition is in the same file as the `styled()` call.

**Comment by mxstbr** [comment]
...which wouldn't work if people use Flow or TypeScript 😅

**Comment by azz** [comment]
True - though not a lot of people run ESLint on their TypeScript, either.

**Comment by James-E-Adams** [comment]
@azz's suggestion may be the way to go then.

**Comment by protonmesh** [comment]
This has bitten me multiple times as well.

The only fool-proof way I see to address this is to inspect the rendered result of the styled component and check whether the `generatedClassName` was actually tacked on to the (topmost?) DOM node. 

There are some concerns with this such as the fact that the extended component might choose to pass the className to a DOM node that is not the topmost.

Also I can't think of any way to reliably get the topmost DOM node other than `ReactDOM.findDOMNode`. I don't think any of solution that uses `refs` would be applicable here when we don't have control over the extended 3rd party component.

The `ReactDOM.findDOMNode` would add a performance hit but this check could be enabled in just dev.

**Comment by chengjianhua** [comment]
Can we implement this by getter?

Whatever the third party component read the `className` explicitly or implicitly, we can check this by setting a flag to be true if the user accesses the getter.

```javascript
const props = {};

Object.defineProperties(props, {
  className: {
    get() {
      console.log('className has been accessed');

      return 'this_is_a_class_name';
    },
    enumerable: true,
  }
})

const { a, b } = props; // no output in console

const { className } = props; // output 'className has been accessed' once

const { ...restProps } = props; // output 'className has been accessed' once


const { className, ...resetProps } = props; // output 'className has been accessed' twice
```

When the third party component using rest props to consume the props, we should treat this as the deliberate behavior to access the `className` property because the third party component would spread the `restProps` on an element most likely. Even though we have a misjudgment, it's okay that this is a warning in development after all.

**Comment by Fer0x** [comment]
@chengjianhua getters or even Proxy doesn't work because React copy all props to new object.

I think the only solution is to use ReactDOM.findDOMNode + some query selectors. It should work for most cases.

**Comment by Fer0x** [comment]
Closed via https://github.com/styled-components/styled-components/commit/fb083e2b7e86153b01e64ca024ab2e66d041aecb (https://github.com/styled-components/styled-components/pull/2073)
